wandb/sdk/internal/job_builder.py

```diff
@@ -0,0 +1,252 @@
+"""
+job builder.
+"""
+import json
+import os
+import sys
+from typing import TYPE_CHECKING, Any, Dict, List, Optional, Tuple, Union
+
+from wandb.sdk.data_types._dtypes import TypeRegistry
+from wandb.sdk.lib.filenames import DIFF_FNAME, METADATA_FNAME, REQUIREMENTS_FNAME
+from wandb.sdk.wandb_artifacts import Artifact
+from wandb.util import make_artifact_name_safe
+
+from .settings_static import SettingsStatic
+
+if sys.version_info >= (3, 8):
+    from typing import TypedDict
+else:
+    from typing_extensions import TypedDict
+
+if TYPE_CHECKING:
+    from wandb.proto.wandb_internal_pb2 import ArtifactRecord
+
+FROZEN_REQUIREMENTS_FNAME = "requirements.frozen.txt"
+JOB_FNAME = "wandb-job.json"
+JOB_ARTIFACT_TYPE = "job"
+
+
+class GitInfo(TypedDict):
+    remote: str
+    commit: str
+
+
+class GitSourceDict(TypedDict):
+    git: GitInfo
+    entrypoint: List[str]
+    args: List[str]
+
+
+class ArtifactSourceDict(TypedDict):
+    artifact: str
+    entrypoint: List[str]
+    args: List[str]
+
+
+class ImageSourceDict(TypedDict):
+    image: str
+    args: List[str]
+
+
+class JobSourceDict(TypedDict, total=False):
+    _version: str
+    source_type: str
+    source: Union[GitSourceDict, ArtifactSourceDict, ImageSourceDict]
+    input_types: Dict[str, Any]
+    output_types: Dict[str, Any]
+    runtime: Optional[str]
+
+
+class ArtifactInfoForJob(TypedDict):
```

Comment:
you could also use dataclasses here for all your classes?

Reply:
For my own edification, why not use TypedDicts?

Reply:
dataclasses are not json serializable and I need to serialize these, I'm going to leave them as type dicts

Reply:
up to you, you can convert it to dict using `asdict` from `dataclasses`.